src/storage/src/render/upsert.rs

```diff
@@ -765,52 +795,104 @@ where
                 input.next().await
             }
         } {
-            match event {
-                AsyncEvent::Data(cap, mut data) if output_cap.time().less_equal(cap.time()) => {
-                    stage_input(
-                        &mut stash,
-                        &mut data,
-                        &input_upper,
-                        &resume_upper,
-                        upsert_config.shrink_upsert_unused_buffers_by_ratio,
-                    );
-                }
-                AsyncEvent::Data(_cap, mut data) => {
-                    stage_input(
-                        &mut stash,
-                        &mut data,
-                        &input_upper,
-                        &resume_upper,
-                        upsert_config.shrink_upsert_unused_buffers_by_ratio,
-                    );
-                }
-                AsyncEvent::Progress(upper) => {
-                    // Ignore progress updates before the `resume_upper`, which is our initial
-                    // capability post-snapshotting.
-                    if PartialOrder::less_than(&upper, &resume_upper) {
-                        continue;
+            // Buffer as many events as possible. This should be bounded, as new data can't be
+            // produced in this worker until we yield to timely.
+            let events = [event]
+                .into_iter()
+                .chain(itertools::unfold(&mut input, |input| {
+                    input.next().now_or_never().flatten()
+                }))
+                .enumerate();
+
+            let mut partial_drain_time = None;
+            for (i, event) in events {
+                match event {
+                    AsyncEvent::Data(cap, mut data) => {
+                        tracing::trace!(
+                            time=?cap.time(),
+                            updates=%data.len(),
+                            "received data in upsert"
+                        );
+                        stage_input(
+                            &mut stash,
+                            &mut data,
+                            &input_upper,
+                            &resume_upper,
+                            upsert_config.shrink_upsert_unused_buffers_by_ratio,
+                        );
+
+                        let event_time = cap.time();
+                        // If the data is at _exactly_ the output frontier, we can preemptively drain it into the state.
+                        // Data within this set events strictly beyond this time are staged as
+                        // normal.
+                        //
+                        // This is a load-bearing optimization, as it is required to avoid buffering
+                        // the entire source snapshot in the `stash`.
+                        if prevent_snapshot_buffering
+                            && output_cap.time().cmp(event_time) == Ordering::Equal
```

Comment:
I feel like I'm missing something but I will comment again anyway
```suggestion
                            && output_cap.time() == event_time
```